feat: add flags to skip owner-referenced resources

enbiyagoral opened the pull request:
## What this PR does / why we need it?

This PR addresses issue where unused resource reports were cluttered with resources that aren't truly orphaned:

- **ReplicaSets owned by Deployments/StatefulSets**: These are kept for rollback history up to `revisionHistoryLimit` and shouldn't be considered orphaned
- **Jobs created by CronJobs**: These are useful for checking logs of last successful/failed runs and are managed by their parent CronJob

**Added Features:**
- `--skip-deployment-replicasets` flag: Filters out ReplicaSets owned by Deployments or StatefulSets using owner reference checking
- `--skip-cronjob-jobs` flag: Filters out Jobs created by CronJobs using owner reference checking

**Usage Examples:**
```bash
# Skip deployment-owned replicasets
kor replicaset --skip-deployment-replicasets

# Skip cronjob-created jobs  
kor job --skip-cronjob-jobs

# Use both flags together
kor all --skip-deployment-replicasets --skip-cronjob-jobs

# Works with multi-resource queries
kor "replicaset,job,configmap" --skip-deployment-replicasets
```

## PR Checklist

- [ ] This PR adds K8s exceptions (false positives)
- [x] This PR adds new code
- [x] This PR includes tests for new/existing code
- [x] This PR adds docs

## GitHub Issue

Closes [#341](https://github.com/yonahd/kor/issues/341)

## Notes for your reviewers

- **Owner Reference Logic**: Used Kubernetes standard owner reference checking to identify parent-child relationships
- **Backward Compatibility**: All existing functionality remains unchanged, new flags are optional
- **Test Coverage**: Added comprehensive tests for both filtering scenarios including edge cases
- **Multi-resource Support**: Both flags work seamlessly with existing multi-resource queries and output formats
- **Real Cluster Testing**: Verified functionality against actual EKS cluster with deployment history and cronjob schedules

enbiyagoral committed:
Add flags to skip owner-referenced resources

- Add --skip-deployment-replicasets flag to filter out ReplicaSets owned by Deployments/StatefulSets
- Add --skip-cronjob-jobs flag to filter out Jobs created by CronJobs
- Implement owner reference checking in ReplicaSet and Job processing
- Add comprehensive test coverage for both filtering features
- Update README.md with new flag documentation

Fixes issue where old replicasets from deployments and cronjob jobs cluttered unused resource reports

yonahd commented:
@enbiyagoral looks really interesting
Do you think we can expand it to all resources to cover
https://github.com/yonahd/kor/discussions/451?

enbiyagoral commented:
> @enbiyagoral looks really interesting Do you think we can expand it to all resources to cover #451?

Sure, I can look into that today

enbiyagoral committed:
Add ownerReferences support and tests for: DaemonSet, ClusterRole, ConfigMap, CRD, Deployment, Job, Pod, PDB, PVC, ReplicaSet, Role, RoleBinding, Secret, Service, StatefulSet, StorageClass, VolumeAttachment

enbiyagoral commented:
Hi @yonahd, Add ownerReferences support and tests for multiple resource types

- Added and improved ownerReferences support and tests for the following resources:
  DaemonSet, ClusterRole, ConfigMap, CRD, Deployment, Job, Pod, PDB, PVC, ReplicaSet, Role, RoleBinding, Secret, Service, StatefulSet, StorageClass, VolumeAttachment.
- Ensured that resources with ownerReferences are properly skipped when the --ignore-owner-references flag is used.
- Updated both process and structured output tests for these resources.

enbiyagoral committed:
feat: Add ownerReferences support for remaining resource types

- Add ownerReferences check and --ignore-owner-references flag support to:
  * HorizontalPodAutoscaler (HPAs)
  * Ingress resources
  * NetworkPolicy resources
  * ServiceAccount resources

- Add comprehensive test coverage for ownerReferences functionality:
  * Test ownerReferences filtering with --ignore-owner-references flag
  * Test proper resource detection when ownerReferences is empty
  * Test correct resource identification when ownerReferences is present
  * Ensure all resource types handle ownerReferences consistently

- Complete ownerReferences support across all resource types in /pkg/kor
- Maintain consistent behavior with existing --ignore-owner-references implementation

enbiyagoral committed:
README.md: Document --ignore-owner-references flag, remove skip flags

enbiyagoral committed:
fix: resolve linter issues with gci formatting

enbiyagoral commented:
> @enbiyagoral looks like the code isnt passing the linter

Now the code passes the linter with no errors. Thanks for the heads up! 🙏

enbiyagoral commented:
> ⚠️ Please install the !['codecov app svg image'](https://private-user-images.githubusercontent.com/152432831/340411587-e90313f4-9d3a-4b63-8b54-cfe14e7ec20d.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI0NDAzNTMsIm5iZiI6MTc1MjQ0MDA1MywicGF0aCI6Ii8xNTI0MzI4MzEvMzQwNDExNTg3LWU5MDMxM2Y0LTlkM2EtNGI2My04YjU0LWNmZTE0ZTdlYzIwZC5zdmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcxM1QyMDU0MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lNTAwODEzNzUyYjc4ODZlZjI0NzE4YzUyODFjNzdmYjc3MmU5MmVjNGFjNThhNGMzZDQ3OGMwOTA4ZTg3ZGNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3LhKEbUmPSwOBUHkpz7h771XtApWzx14kg8cuF3uCFI) to ensure uploads and comments are reliably processed by Codecov.
> 
> ## [Codecov](https://app.codecov.io/gh/yonahd/kor/pull/467?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen) Report
> Attention: Patch coverage is `83.05085%` with `10 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 44.16%. Comparing base [(`08aef32`)](https://app.codecov.io/gh/yonahd/kor/commit/08aef327a9af028969238fbdc49b838cf5d95367?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen) to head [(`f23a450`)](https://app.codecov.io/gh/yonahd/kor/commit/f23a45007dc78bdf1de88911348b64f5593914f5?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen).
> > Report is 5 commits behind head on main.
> 
> [Files with missing lines](https://app.codecov.io/gh/yonahd/kor/pull/467?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen)	Patch %	Lines
> [pkg/kor/crds.go](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&filepath=pkg%2Fkor%2Fcrds.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen#diff-cGtnL2tvci9jcmRzLmdv)	0.00%	[3 Missing ⚠️ ](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen)
> [pkg/kor/pv.go](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&filepath=pkg%2Fkor%2Fpv.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen#diff-cGtnL2tvci9wdi5nbw==)	33.33%	[1 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen)
> [pkg/kor/rolebindings.go](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&filepath=pkg%2Fkor%2Frolebindings.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen#diff-cGtnL2tvci9yb2xlYmluZGluZ3MuZ28=)	33.33%	[1 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen)
> [pkg/kor/statefulsets.go](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&filepath=pkg%2Fkor%2Fstatefulsets.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen#diff-cGtnL2tvci9zdGF0ZWZ1bHNldHMuZ28=)	33.33%	[1 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen)
> [cmd/kor/root.go](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&filepath=cmd%2Fkor%2Froot.go&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen#diff-Y21kL2tvci9yb290Lmdv)	0.00%	[1 Missing ⚠️ ](https://app.codecov.io/gh/yonahd/kor/pull/467?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen)
> ❗ Your organization needs to install the [Codecov GitHub app](https://github.com/apps/codecov/installations/select_target) to enable full functionality.
> 
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/yonahd/kor/pull/467?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Yonah+Dissen).
> 
> 🚀 New features to boost your workflow:

83% patch coverage looks good!

yonahd reviewed (approved):
Lgtm

enbiyagoral commented:
Thanks for the review and merge! 🙌 @yonahd

yonahd commented:
> Thanks for the review and merge! 🙌 @yonahd 

Thanks for the contribution and great work!
Hoping to see you involved more in the project 🙂
